parlai/agents/transformer/polyencoder.py

```diff
@@ -267,16 +289,19 @@ def encode(self, ctxt_tokens, cand_tokens):
             assert len(ctxt_tokens.shape) == 2
             bsz = ctxt_tokens.size(0)
             # get context_representation. Now that depends on the cases.
-            ctxt_out, ctxt_mask = self.encoder_ctxt(ctxt_tokens)
+            ctxt_out, ctxt_mask, ctxt_pos = self.encoder_ctxt(ctxt_tokens)
+            att_keys = ctxt_out if self.attention_keys == 'context' else ctxt_pos
             dim = ctxt_out.size(2)
 
             if self.type == 'codes':
                 ctxt_rep = self.attend(
                     self.code_attention,
-                    self.codes.repeat(bsz, 1, 1),
-                    ctxt_out,
-                    ctxt_mask,
+                    queries=self.codes.repeat(bsz, 1, 1),
+                    keys=att_keys,
+                    values=ctxt_out,
+                    mask=ctxt_mask,
                 )
+                ctxt_pos = None  # we don't need this anymore
```

Comment:
why don't we need this anymore?

Reply:
ctxt position embs are used as attention keys only when the values are the transformer outputs. for the `codes` case, we do not need them for the secondary attention over the `codes`